### ISSUE TITLE
IfThenElse Typing and further questions

### ISSUE THREAD

**Issue by samysweb** [dlshriver/dnnv#75]
Hello,
first of all thank you very much for developing DNNV!
While I'm still struggling a bit with getting things to work the way I want (see below), this type of tool is very helpful for me at the moment as I am trying to verify more complex (safety) properties on neural network control systems (multiple conjunctions, disjunctions, maxima/minima etc.).

As I was working on automatically generating properties to check, I stumbled across the (in my eyes) curious typing behaviour of the `IfThenElse` Operator and was hoping you might be able to help me.
For context: I am mainly trying to use the `IfThenElse` Operator to generate `max` or `min` functions.

Note, that I'm filing this as an issue, because I believe that it's a bug.
If it turns out not to be a bug let me know and I'll ask my question in the Discussion board instead.

## Problem 1: Encapsulated `IfThenElse`
Consider the following property (I tried to boil the property down somewhat to make it readable, this is not the property I want to verify eventually):
```python
from dnnv.properties import *
import numpy as np

N = Network("N")

x_min = np.array([[0, -200]])
x_max = np.array([[100, 100]])
Forall(
	x_,
	Implies(
		And(
			And(x_min <= x_, x_ <= x_max),
			0.0 >= (
				2.0 * 
					IfThenElse(
						0.0>IfThenElse(
								(0.1 * (x_[0,1] - -210.0))<=(-0.2 * (x_[0,1] - 100)),
								(0.1 * (x_[0,1] - -210.0)),
								(-0.2 * (x_[0,1] - 100))),
						0.0,
						IfThenElse(
							(0.1 * (x_[0,1] - -210.0))<=(-0.2 * (x_[0,1] - 100)),
							(0.1 * (x_[0,1] - -210.0)),
							(-0.2 * (x_[0,1] - 100)))
					)
			)
		),
		And((N(x_) >= -100.0), (N(x_) < 100.0))
	)
)
```
I believe the important part are the encapsulated `IfThenElse` blocks.
When trying to execute this, I get an error: `DNNVTypeError(Incompatible types in And)`.
However, when I remove the `2.0 *` inside the brackets on the right side of `>=` the property compiles fine.

## Problem 2: Types in `IfThenElse` branches
I also encountered a similar problem on a few occasions where one of the two `IfThenElse` branches was a scalar while the otherone uses a variable (again not a property I actually want to check, but a simplification):
```python
from dnnv.properties import *
import numpy as np

N = Network("N")

x_min = np.array([[0, -200]])
x_max = np.array([[100, 100]])
Forall(
	x_,
	Implies(
		And(x_min <= x_, x_ <= x_max),
		Or(
			Or(
				(N(x_) == 100.0),
				And(
					(0.0 >= IfThenElse(0.0>x_[0,1],0.0,x_[0,1])),
					And(N(x_) >= -100.0, N(x_) < 100.0)
				)
			),
			And(
				x_[0,1] == 0.0,
				N(x_) == 0.0
			)
		)
	)
)
```
This returned `DNNVTypeError(Incompatible types in IfThenElse)`.
Taking into account the simplification of `0.0*` and the nature of floating points, I was hoping one might be able to substitute the scalar `0.0` with something like `(0.16*x_[0,0]+ -0.16*x_[0,0])`, however this lead to another error:
```
Traceback (most recent call last):
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/bin/dnnv", line 8, in <module>
    sys.exit(_main())
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/__main__.py", line 115, in _main
    return exit(main(*cli.parse_args()))
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/__main__.py", line 77, in main
    result, cex = verifier.verify(phi, **params)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/verifiers/common/base.py", line 84, in verify
    return cls(phi, **kwargs).run()
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/verifiers/common/base.py", line 112, in run
    for subproperty in self.reduce_property():
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/verifiers/common/base.py", line 96, in reduce_property
    for subproperty in self.reduction().reduce_property(~self.property):
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/verifiers/common/reductions/iopolytope/reduction.py", line 72, in reduce_property
    dnf_expression = expression.canonical()
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/expressions/base.py", line 162, in canonical
    return CanonicalTransformer().visit(self)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 146, in visit_Or
    expr = self.visit(expr)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 35, in visit_And
    expr = self.visit(expr)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 66, in visit_Exists
    expr = self.visit(expression.expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 35, in visit_And
    expr = self.visit(expr)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/canonical.py", line 66, in visit_Equal
    return self.visit(Or(And(expr1 <= expr2, expr2 <= expr1)))
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 146, in visit_Or
    expr = self.visit(expr)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 35, in visit_And
    expr = self.visit(expr)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/canonical.py", line 91, in visit_LessThanOrEqual
    self.visit(Add(expression.expr1, Multiply(Constant(-1), expression.expr2)))
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/canonical.py", line 54, in visit_Add
    coefficients = self._extract_coefficients(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/canonical.py", line 27, in _extract_coefficients
    expr = self.visit(expr)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/dnf.py", line 28, in visit
    expression = super().visit(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/visitors/base.py", line 19, in visit
    self.visited[expression] = visitor(expression)
  File "/home/steuber/Dokumente/Hortus/Univ/KIT/Master/Work/repos/DNNV/.venv/lib/python3.7/site-packages/dnnv/properties/transformers/canonical.py", line 114, in visit_Multiply
    for e in expressions[0]:
IndexError: list index out of range
```

## Summary
In conclusion it would be very helpful if you could clarify to what extend the errors above are to be expected or whether they are bugs (i.e. am I doing something wrong or is DNNV supposed to behave differently).

**For replication purposes:**
I'm using the most up-to-date version of DNNV (tag `v0.5.1`) with a virtual environment as described in the README.
If you need any more information to replicate the issue, please let me know.

Thanks in advance,
Samuel

**Comment by dlshriver** [comment]
I believe your understanding of IfThenElse is correct. This definitely looks like a bug. v0.5 added some new type checking code for property specifications, and it looks like the type constraints it is inferring for the IfThenElse branches are too tight, making them incompatible. I'll take a closer look.

**Comment by samysweb** [comment]
Thanks a lot for the quick response!
I've just tried rerunning my examples and they seem to be working fine using the `fix-ifthenelse` branch.
